### PR TITLE
EPMLABSBRN-822 show exercises

### PR DESCRIPTION
**Pull request by alextanlan**
[EPMLABSBRN-822](https://jira.epam.com/jira/browse/EPMLABSBRN-822)

**Description**:

Implemented (partially) an exercises page:

- added the exercises module with routing

- created selectors Groups/Series/Subgroups/Exercises (with some pre-polluted values by default)

- added an exercises table

image

P.S. There will be additional PR with polishing details.

### CODE REVIEW

File: frontend-angular/src/app/modules/admin/admin.module.ts [hunk 1]
```diff
@@ -1,17 +1,27 @@
 import { NgModule } from '@angular/core';
 import { CommonModule } from '@angular/common';
 import { MatToolbarModule } from '@angular/material/toolbar';
+import { MatButtonModule } from '@angular/material/button';
+
```
Comment (KomlevDm) [inline frontend-angular/src/app/modules/admin/admin.module.ts:5]: empty line

File: frontend-angular/src/app/modules/admin/models/exercise.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
+export interface Exercise {
```
Comment (KomlevDm) [inline frontend-angular/src/app/modules/admin/models/exercise.ts:1]: 1) good practice for naming of interfaces with prefix 'I' -> IExercise
2) multiple entities in one file - bad practice
3) better all entities create as classes for using class-transformer lib in the future and manipulate real class and not raw objects (this is a good way for this app, there are no high performance requirements, and creating a class is cheap)

File: frontend-angular/src/app/modules/admin/modules/exercises/components/exercises/exercises.component.scss [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    padding: 40px $header-horizontal-padding;
+  }
+
+  &__table {
+    width: 100%;
+    box-shadow: none;
+  }
+}
+
+::ng-deep {
```
Comment (KomlevDm) [inline frontend-angular/src/app/modules/admin/modules/exercises/components/exercises/exercises.component.scss:20]: need use :host

File: frontend-angular/src/app/modules/admin/modules/exercises/components/exercises/exercises.component.spec.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
+import { ExercisesComponent } from './exercises.component';
+
+describe('ExercisesComponent', () => {
+  let component: ExercisesComponent;
+  let exercisesApiServiceMock;
+  let changeDetectorRefMock;
+
+  beforeEach(async () => {
```
Comment (KomlevDm) [inline frontend-angular/src/app/modules/admin/modules/exercises/components/exercises/exercises.component.spec.ts:8]: 1) Why are you use async?
2) What is a strange approach for testing?

File: frontend-angular/src/app/modules/admin/modules/exercises/components/exercises/exercises.component.ts [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+  dataSource: MatTableDataSource<any>;
+
+  seriesName$ = new BehaviorSubject<string>('');
+  private groupId$ = new Subject<string>();
+  private subGroupId$ = new Subject<string>();
+  private subscription: Subscription;
+  private readonly LOG_SOURCE = 'ExercisesComponent';
+
+  constructor(
+    private exercisesApiService: ExercisesApiService,
```
Comment (KomlevDm) [inline frontend-angular/src/app/modules/admin/modules/exercises/components/exercises/exercises.component.ts:40]: nit: readonly

File: frontend-angular/src/app/modules/admin/modules/exercises/components/exercises/exercises.component.ts [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+  onSeriesChange(seriesName: string): void {
+    this.seriesName$.next(seriesName);
+  }
+
+  onSubGroupChange(subGroupId: string): void {
+    this.subGroupId$.next(subGroupId);
+  }
+
+  ngOnDestroy(): void {
+    this.subscription.unsubscribe();
```
Comment (KomlevDm) [inline frontend-angular/src/app/modules/admin/modules/exercises/components/exercises/exercises.component.ts:63]: destroyer$ approach is more universal for scaling

File: frontend-angular/src/app/modules/admin/modules/exercises/components/exercises/exercises.component.ts [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+  }
+
+  isEnableChanged(exercise, isEnable: boolean) {
```
Comment (KomlevDm) [inline frontend-angular/src/app/modules/admin/modules/exercises/components/exercises/exercises.component.ts:66]: different styles of function signature - good practice to always use access modifier and return type

File: frontend-angular/src/app/modules/admin/modules/exercises/components/exercises/exercises.component.ts [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+    console.log('enable changed:', exercise, isEnable);
+  }
+
+  // TODO: implement a real logic for task matrix (for now it is a draft)
+  getMatrixFromTasks(tasks: Task[]): string {
+    if (!tasks || tasks.length === 0) {
```
Comment (KomlevDm) [inline frontend-angular/src/app/modules/admin/modules/exercises/components/exercises/exercises.component.ts:72]: !tasks?.length

File: frontend-angular/src/app/modules/admin/modules/exercises/components/exercises/exercises.component.ts [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+      return '';
+    }
+
+    let res = '';
+    tasks.forEach((task: Task) => {
```
Comment (KomlevDm) [inline frontend-angular/src/app/modules/admin/modules/exercises/components/exercises/exercises.component.ts:77]: use reduce and can delete extra variable

File: frontend-angular/src/app/modules/admin/modules/exercises/components/exercises/exercises.component.ts [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+      const row  = task.answerOptions.map((answer: Answer) => answer.word).join(' ');
+      res += row + '\n';
+    });
+
+    return res;
+  }
+
+  private initDataSource(): void {
```
Comment (KomlevDm) [inline frontend-angular/src/app/modules/admin/modules/exercises/components/exercises/exercises.component.ts:85]: this good signature and full👍

File: frontend-angular/src/app/modules/admin/modules/exercises/components/exercises/exercises.component.ts [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+    this.displayedColumns = [ 'id', 'seriesId', 'name', 'level', 'noise', 'noiseSound', 'tasks', 'available'];
+  }
+
+  private initExercises() {
+    this.subscription = combineLatest([this.groupId$, this.seriesName$, this.subGroupId$]).pipe(
+      map((argsArray: string[]) => {
+        const allIdsExist = argsArray.every(x => !!x);
```
Comment (KomlevDm) [inline frontend-angular/src/app/modules/admin/modules/exercises/components/exercises/exercises.component.ts:92]: nit: argsArray.every(Boolean)

File: frontend-angular/src/app/modules/admin/modules/exercises/components/exercises/exercises.component.ts [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+        if (!allIdsExist) {
+          this.hideExercisesTable();
+        }
+        return allIdsExist ? argsArray[2] : false;
```
Comment (KomlevDm) [inline frontend-angular/src/app/modules/admin/modules/exercises/components/exercises/exercises.component.ts:96]: what is a magic value === 2?

File: frontend-angular/src/app/modules/admin/modules/exercises/components/exercises/exercises.component.ts [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+      }),
+      filter(Boolean),
```
Comment (KomlevDm) [inline frontend-angular/src/app/modules/admin/modules/exercises/components/exercises/exercises.component.ts:98]: filter<string>(Boolean) - and don't have to use switchMap((subGroupId: string) - typescript will do it

File: frontend-angular/src/app/modules/admin/modules/exercises/components/exercises/select-panel/select-panel.component.ts [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+    this.ngUnsubscribe.next();
+    this.ngUnsubscribe.complete();
+  }
+
+  private initGroups(): void {
+    this.groups$ = this.languagesControl.valueChanges.pipe(
+      startWith(DEFAULT_SELECT_VALUE.languageId),
+      tap((languageId: string) => {
+        this.groupsControl.reset();
+        return languageId;
```
Comment (KomlevDm) [inline frontend-angular/src/app/modules/admin/modules/exercises/components/exercises/select-panel/select-panel.component.ts:87]: what for? operator tap does not participate in chain

File: frontend-angular/src/app/modules/admin/modules/exercises/components/exercises/select-panel/select-panel.component.ts [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+    );
+  }
+
+  private initSubGroups(): void {
+    this.subGroups$ = this.seriesControl.valueChanges.pipe(
+      startWith(DEFAULT_SELECT_VALUE.seriesIdName),
+      tap((seriesIdAndName: string) => {
+        this.subGroupsControl.reset();
+        if (seriesIdAndName) {
+          const seriesName = seriesIdAndName.split(';')[1];
```
Comment (KomlevDm) [inline frontend-angular/src/app/modules/admin/modules/exercises/components/exercises/select-panel/select-panel.component.ts:129]: why 1 (magic value) - better create const with good name

File: frontend-angular/src/app/modules/admin/modules/exercises/components/exercises/select-panel/select-panel.component.ts [hunk 16]
```diff
@@ -0,0 +1,181 @@
...
+          this.seriesChanged.emit(seriesName);
+        }
+      }),
+      switchMap((seriesIdAndName: string) => {
+        const seriesId = seriesIdAndName ? seriesIdAndName.split(';')[0] : null;
```
Comment (KomlevDm) [inline frontend-angular/src/app/modules/admin/modules/exercises/components/exercises/select-panel/select-panel.component.ts:134]: why 0 (magic value) - better create const with good name

File: frontend-angular/src/app/modules/admin/modules/exercises/components/exercises/select-panel/select-panel.component.ts [hunk 17]
```diff
@@ -0,0 +1,181 @@
...
+      takeUntil(this.ngUnsubscribe)
+    ).subscribe();
+  }
+
+  private setDefaultSelectValues() {
+    setTimeout(() => {
+      this.setDefaultLanguage();
+      this.setDefaultGroup();
+      this.setDefaultSeries();
+    }, 0);
```
Comment (KomlevDm) [inline frontend-angular/src/app/modules/admin/modules/exercises/components/exercises/select-panel/select-panel.component.ts:167]: what is it? Look like a hack

try use AfterViewInit fo it

File: frontend-angular/src/app/modules/admin/modules/exercises/exercises.module.ts [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+
+@NgModule({
+  declarations: [ExercisesComponent, SelectPanelComponent],
+  imports: [
+    CommonModule,
+    ExercisesRoutingModule,
+    ReactiveFormsModule,
+    MatFormFieldModule,
+    MatSelectModule,
+    MatIconModule,
```
Comment (KomlevDm) [inline frontend-angular/src/app/modules/admin/modules/exercises/exercises.module.ts:23]: I don't see <mat-icon> in this PR

File: frontend-angular/src/app/modules/admin/modules/exercises/components/exercises/exercises.component.scss (hunk 3 above)
Comment (lifeart) [reply frontend-angular/src/app/modules/admin/modules/exercises/components/exercises/exercises.component.scss:20]: @KomlevDm let's pospone it due to scss linting rule in ci

File: frontend-angular/src/app/modules/admin/models/exercise.ts (hunk 2 above)
Comment (lifeart) [reply frontend-angular/src/app/modules/admin/models/exercise.ts:1]: interface naming should be part of CI

File: frontend-angular/src/app/modules/admin/admin.module.ts (hunk 1 above)
Comment (lifeart) [reply frontend-angular/src/app/modules/admin/admin.module.ts:5]: should be part of CI

File: frontend-angular/src/app/modules/admin/modules/exercises/components/exercises/select-panel/select-panel.component.ts (hunk 15 above)
Comment (lifeart) [reply frontend-angular/src/app/modules/admin/modules/exercises/components/exercises/select-panel/select-panel.component.ts:129]: let's adress it for next MR

File: frontend-angular/src/app/modules/admin/modules/exercises/components/exercises/select-panel/select-panel.component.ts (hunk 16 above)
Comment (lifeart) [reply frontend-angular/src/app/modules/admin/modules/exercises/components/exercises/select-panel/select-panel.component.ts:134]: let's adress it for next MR

Comment (lifeart) [reply frontend-angular/src/app/modules/admin/modules/exercises/components/exercises/select-panel/select-panel.component.ts:134]: let's adress it for next MR

File: frontend-angular/src/app/modules/admin/modules/exercises/components/exercises/select-panel/select-panel.component.ts (hunk 15 above)
Comment (lifeart) [reply frontend-angular/src/app/modules/admin/modules/exercises/components/exercises/select-panel/select-panel.component.ts:129]: let's adress it for next MR

**Review by lifeart — APPROVED** [review]

**Comment by lifeart** [comment]
@alextanlan thanks for your work! Going to merge current request, let's adress non-automatable comments in follow-up MR

**Pull request merged**